### ISSUE TITLE
Limit for sinv needs to be increased

### ISSUE THREAD

**Issue by SudhirNadiga-NOAA** [NOAA-EMC/NCEPLIBS-bufr#579]
When running sinv  on a large tank, I get an error message.

[clogin07 /lfs/h2/emc/obsproc/noscrub/sudhir.nadiga]$ ls -l /lfs/h1/ops/para/dcom/20240318/b021/xx054
-rw-rw-r-- 1 ops.para para 2073922960 Mar 19 01:16 /lfs/h1/ops/para/dcom/20240318/b021/xx054
[clogin07 /lfs/h2/emc/obsproc/noscrub/sudhir.nadiga]$ 

**ERROR MESSAGE BELOW**
[clogin07 /lfs/h2/emc/obsproc/noscrub/sudhir.nadiga]$ sinv /lfs/h1/ops/para/dcom/20240318/b021/xx054
 +++++++++++++++++++++WARNING+++++++++++++++++++++++
 BUFRLIB: UFBTAB - THE NO. OF DATA SUBSETS IN THE BUFR FILE IS .GT. LIMIT OF 16000000 IN THE 4TH ARG. (INPUT) - INCOMPLETE READ  
 >>>UFBTAB STORED 15999291 REPORTS OUT OF ********<<<                                                                            
 +++++++++++++++++++++WARNING+++++++++++++++++++++++
  
 +++++++++++++++++++++WARNING+++++++++++++++++++++++
 BUFRLIB: UFBTAB - THE NO. OF DATA SUBSETS IN THE BUFR FILE IS .GT. LIMIT OF 16000000 IN THE 4TH ARG. (INPUT) - INCOMPLETE READ  
 >>>UFBTAB STORED 15999291 REPORTS OUT OF ********<<<                                                                            
 +++++++++++++++++++++WARNING+++++++++++++++++++++++
  
 
209  NOAA 18              9109137  000                                                                                
223  NOAA 19              6890154  000                                                                                

                         15999291

[clogin07 /lfs/h2/emc/obsproc/noscrub/sudhir.nadiga]$ 

How do we address this issue? Thanks.

**Comment by jbathegit** [comment]
This is a parameter setting in the sinv utility, based on the expected maximum number of data subsets that one would ever expect to read from a single BUFR file.  We could certainly set it to a larger number, but since we're already at 16 million, and since that number is used to dimension two underlying real*8 arrays, then at some point we could conceivably reach a limit where the resulting compiled object is too big to load into RAM.  So we may also need to modify the utility to redefine the underlying said and siid arrays as allocatable and dynamically allocate them at run time, rather than fixing their sizes at compile time.

Either way, we'd need to set some practical limit in the utility.  @SudhirNadiga-NOAA do you have any idea how much larger you'd need this setting to be?  Note that you can get the count of subsets in any BUFR file by just calling [ufbtab](https://noaa-emc.github.io/NCEPLIBS-bufr/ufbtab_8f.html) with a negative logical unit number, and if you do that then you don't need to worry about how big your array is actually dimensioned because it won't actually try to read and store any of the requested mnemonics in the last argument.

Also CCing @jack-woollen for his awareness :-)

**Comment by jbathegit** [comment]
FWIW, I just did this for the file you mentioned above, and it had just under 110 million subsets in it!  Do you know if that's a typical daily count for these files?

**Comment by SudhirNadiga-NOAA** [comment]
Thanks for all your efforts. I will need to ask Iliana if she knows how big our tanks can get in the next year or so. My guess is that  the biggest tanks we have are the b021/xx206 tanks.
In the past, the poes-sst tanks used to be our biggest tanks, but these CrIS 431 tanks are huge, and the CrIS 2211 tanks (only made by development) are even bigger.  I don't know if that fully answers your question, but I can ask Iliana if she has a better answer.
[clogin06 /lfs/h1/ops/prod/dcom/20240319]$ ls -lrt b021/*
-rw-rw-r-- 1 dfprod   prod  263835224 Mar 20 00:05 b021/xx042
-rw-rw-r-- 1 dfprod   prod  349471088 Mar 20 00:12 b021/xx213
-rw-rw-r-- 1 ops.prod prod  413967832 Mar 20 00:21 b021/xx246
-rw-rw-r-- 1 dfprod   prod 1621221504 Mar 20 00:32 b021/xx039
-rw-rw-r-- 1 ops.prod prod  603775608 Mar 20 00:35 b021/xx046
-rw-rw-r-- 1 ops.prod prod 1951418592 Mar 20 00:36 b021/xx045
-rw-rw-r-- 1 dfprod   prod   55290464 Mar 20 00:37 b021/xx044
-rw-rw-r-- 1 ops.prod prod 1698900976 Mar 20 01:04 b021/xx248
-rw-rw-r-- 1 dfprod   prod 1346407584 Mar 20 01:13 b021/xx239
-rw-rw-r-- 1 dfprod   prod  235666528 Mar 20 01:13 b021/xx036
-rw-rw-r-- 1 dfprod   prod   66225472 Mar 20 01:13 b021/xx033
-rw-rw-r-- 1 ops.prod prod  947068136 Mar 20 01:16 b021/xx053
-rw-rw-r-- 1 ops.prod prod 1976323824 Mar 20 01:16 b021/xx054
-rw-rw-r-- 1 ops.prod prod   62270376 Mar 20 01:17 b021/xx028
-rw-rw-r-- 1 dfprod   prod  140132320 Mar 20 01:19 b021/xx035
-rw-rw-r-- 1 ops.prod prod 5941454184 Mar 20 01:34 b021/xx206
-rw-rw-r-- 1 ops.prod prod  828036152 Mar 20 01:51 b021/xx051
-rw-rw-r-- 1 ops.prod prod 1971780304 Mar 20 01:52 b021/xx052
-rw-rw-r-- 1 ops.prod prod 3153255264 Mar 20 01:53 b021/xx241
-rw-rw-r-- 1 ops.prod prod  350048312 Mar 20 01:55 b021/xx201
-rw-rw-r-- 1 ops.prod prod   56303488 Mar 20 01:59 b021/xx023
-rw-rw-r-- 1 ops.prod prod   56355784 Mar 20 01:59 b021/xx123
-rw-rw-r-- 1 ops.prod prod  185883912 Mar 20 02:00 b021/xx027
-rw-rw-r-- 1 ops.prod prod  870942000 Mar 20 08:19 b021/xx203
-rw-rw-r-- 1 dfprod   prod  307493728 Mar 20 08:48 b021/xx038
-rw-rw-r-- 1 dfprod   prod 2338872816 Mar 20 12:14 b021/xx212
[clogin06 /lfs/h1/ops/prod/dcom/20240319]$ uftab b021/xx042

**Comment by SudhirNadiga-NOAA** [comment]
@ilianagenkova Do we have an idea as to the maximum desired for sinv in terms of number of subsets?

**Comment by SudhirNadiga-NOAA** [comment]
It looks like the xx054 tank may have more subsets by  a lot (~18X )
[clogin07 /lfs/h1/ops/prod/dcom/20240319/b021]$ binv xx206
 
type        messages       subsets         bytes
 
NC021206       32409       5832000    1646306317      179.95
TOTAL          32409       5832000    1646306317
 
[clogin07 /lfs/h1/ops/prod/dcom/20240319/b021]$ binv xx054
 
type        messages       subsets         bytes
 
NC021054     4943114     104542051    1953387584       21.15
TOTAL        4943114     104542051    1953387584
 
[clogin07 /lfs/h1/ops/prod/dcom/20240319/b021]$

**Comment by SudhirNadiga-NOAA** [comment]
I am running binv on all our tanks, so we should have an idea as to the largest tanks by number of subsets. This will help answer your question.

**Comment by SudhirNadiga-NOAA** [comment]
My check shows that the poes_sst files have close to 600 million subsets. 
 
type        messages       subsets         bytes
 
NC012023      139362     556549305    -990011533     3993.55
TOTAL         139362     556549305    -990011533

Iliana and I will discuss this problem and get back to you on any further details, since we are considering writing the two different instruments in different tanks to reduce the number of subsets and for ease of dumping.

**Comment by jbathegit** [comment]
@SudhirNadiga-NOAA @ilianagenkova any updates on this?

**Comment by ilianagenkova** [comment]
I'll provide the largest obs counts that we see (per tank) in the next few days.

**Comment by jbathegit** [comment]
@SudhirNadiga-NOAA @ilianagenkova any updates?  I'm asking b/c I never heard anything more from either of you about this.

Or, if you're no longer interested in pursuing this, then please feel free to close this issue.

**Comment by ilianagenkova** [comment]
@jbathegit , the largest tanks at the moment are ~10GB
ls -l /lfs/h1/ops/prod/dcom//20241113/b012/xx023                                                                                                                                    
-rw-rw-r-- 1 ops.prod prod **10711038888** Nov 14 01:59 /lfs/h1/ops/prod/dcom//20241113/b012/xx023

and they contain about ~700-800 millions subsets:

[clogin03 /lfs/h2/emc/obsproc/noscrub/iliana.genkova/TEMPTEST]$ binv  /lfs/h1/ops/prod/dcom//20241113/b012/xx023                                                                                                                                   
type        messages       subsets         bytes 
NC012023      195735     781677415    2119893694     3993.55
TOTAL         195735     781677415    2119893694

**sinv** already can't handle these. If it would be too much work to update sinv, we can use the simple code we wrote:
/lfs/h2/emc/obsproc/noscrub/iliana.genkova/Utils/READALL/count_readall.x (on Cactus)

**Comment by jack-woollen** [comment]
@ilianagenkova The problem with sinv as it is, is it tries to reads all the data into a table first, which can be quicker than parsing each subset separately. I didn't envision multiple hundreds of million subsets in a file when writing it! I could rewrite it (or a lookalike code) to just read through and count the data which doesn't require storing it. That's pretty simple. I'll make a sample  to try and see if that works out.

**Comment by ilianagenkova** [comment]
Again, only if it's relatively easy lift. I don't know how many ppl use the NCEPLIBS-bufr utilities and if it's justified to update it only for us/obsproc.   My count_readall.x is pretty slow but at least it does the job.

**Comment by jack-woollen** [comment]
@ilianagenkova My reading everything one at a time is no faster than yours I'm sure. There is some tricky things that might work faster, but will take longer to develop. I'll let you know if I find the time to find them! Its an interesting problem for sure.

**Comment by jack-woollen** [comment]
@ilianagenkova on hera:

![image](https://github.com/user-attachments/assets/c53bc2cb-5dd1-4c83-8085-dbb51d7783fd)

I'll make one on wcoss2.

**Comment by jack-woollen** [comment]
@ilianagenkova  - the wcoss2 version is here - runt builds sinv.x and runs a timing test - wcoss2 is a little slower than hera.

![image](https://github.com/user-attachments/assets/3da6b595-6caa-4c73-bf72-efcc46848efe)

**Comment by ilianagenkova** [comment]
@jack-woollen, this is fantastic! Thank you!
I tested the code on wcoss2, it works for me and it runs fast - the new sinv handled a 11GB tank (one of the largest I've seen) in less about a minute. I'll copy the code so I have a copy of it, and will study it when I get a moment. 

I will leave it up to you and @jbathegit  to decide whether to add the new sinv to the Bufr library utilities as one more utility, or to replace the current sinv. Also, you can do it when it's convenient for you. We can use your executable until then.

@SudhirNadiga-NOAA @SteveStegall-NOAA , you can test it too (Dogwood):
/lfs/h2/emc/global/noscrub/Jack.Woollen/jwfork/newsinv/sinvx
or 
 /lfs/h2/emc/obsproc/noscrub/iliana.genkova/Utils/JW_newsinv/sinv.x (

**Comment by jbathegit** [comment]
@jack-woollen thanks for this, but it looks like whatever changes you made in `/lfs/h2/emc/global/noscrub/Jack.Woollen/jwfork/newsinv` were to a very old version of the sinv code, rather than to the [latest release version in the develop branch](https://github.com/NOAA-EMC/NCEPLIBS-bufr/blob/develop/utils/sinv.F90.in) of the NCEPLIBS-bufr GitHub repository.  For example:

- your code is still the older F77, whereas the latest release version is now F90
- your code doesn't have the recent update to allow specification of the master table directory as an optional 2nd call argument
- and most significantly, your code still has the old block sadline and sidline arrays, which means all of the satellite and instrument meanings are still hardcoded into the sinv source code rather than using [getcfmng()](https://noaa-emc.github.io/NCEPLIBS-bufr/cftbvs_8F90.html#af113e7ee7e2023cd466c2b28616cd64a) to read them from the master tables

Is there some particular reason you made your updates to an older version of the sinv code, rather than to the latest release version in the develop branch?  I'm asking b/c I don't see how we can merge any of this into the develop branch in GitHub, unless that was never your intent to begin with(?)

**Comment by jack-woollen** [comment]
@jbathegit Good point. I never meant to merge this code into develop because of the significant change to ufbtab that is needed along with it. I have to think that through when I have more time. I'll get back to it when I have a chance to work it through.

**Comment by ilianagenkova** [comment]
@jack-woollen , are the changes in ufbtab to speeding it up or to handle larger data arrays?

**Comment by jack-woollen** [comment]
@ilianagenkova The change to ufbtab allows an app to call it repeatedly to extract data progressively from one input file. The current ufbtab is one and done. If it can't read all the data in one call, it errs off. Since the modified sinv is collapsing the data by itemizing counts of satellite types, it doesn't need to store all the data at one time. The purpose of ufbtab has always been to turbo read simple elements from complicated files. That doesn't change. But the modified version defines output arguments differently. Probably I will introduce a new version of ufbtab with a new name to preserve the current interface of ufbtab. When I get a couple more days to work on it, I'll merge the new stuff into develop. Other things are calling at the moment!

**Comment by ilianagenkova** [comment]
@jack-woollen , would your workload allow you to update the new/faster _ufbtab_  (following @jbathegit comment  earlier in the tread, from Nov 26, 2024)? 
If it can be added (as ufbtab2 perhaps) to a next bufr update, we can use it to speed up different calls to it in bufr-dump. 
thanks!

**Comment by jack-woollen** [comment]
@ilianagenkova I will. Thanks for reminding me.

**Comment by jack-woollen** [comment]
@jbathegit @ilianagenkova I just pushed a branch to NCEPLIBS_bufr called jsw-sinv-ufbtab to settle this issue. Jeff, if you do a pre-PR review we can get caught up on this and maybe simplify the PR. I'm not sure if this needs new test or not but something to consider.

**Comment by ilianagenkova** [comment]
Tagging @dmerkova , she will test obsproc with the new UFBTAB

**Comment by jbathegit** [comment]
Thanks for your work on this Jack, and I really like your approach to use an internal switch to toggle on/off the reading of data in chunks so that it can be done using the existing ufbtab(), rather than having to create a separate new ufbtab2() routine.  As the code manager for this library, I certainly appreciate minimizing subprogram proliferation where/whenever we can! :-)

My only concern with your overall solution is in how you implemented that toggle switch.  At the end of the day, the sinv utility is an application program, and from a software architecture standpoint application programs should never directly connect to private modules within a library.  In fact, the only NCEPLIBS-bufr module that any Fortran application program should ever need to access is bufr_interface, and only if it's calling C functions within the library.

So instead of giving the sinv utility direct access to the internal modv_vars module, my suggestion would be to toggle the new part variable to .true. by creating a new library subprogram that you can call from sinv to do the trick.  Think of it as similar to how application programs now call setim8b() to toggle the internal im8b variable from .false. to .true., rather than needing direct access to the same internal modv_vars module to toggle that variable.

In fact, you could use setim8b() as a template for your new subprogram (call it, say, setpart()?), and it can be just as short and simple as setim8b().  In fact, and assuming that the plan is for the sinv utility to now always call that new subprogram, then I think the level of testing we currently have available in the test/test_scripts/test_sinv.sh harness will probably be sufficient.  Either way we can always confirm for sure by checking the gcovr output once the PR gets submitted :-)

**Comment by jack-woollen** [comment]
Jeff, I appreciate the need to keep a  lid on the number of subroutines! How about we just put the part switch inside ufbtab.

!  entry to set chunking on or off 
!  -------------------------------

      if(str(1:6)=='partab') then
         if(iret=0) part=.false.
         if(iret=1) part=.true.
         return
      endif

**Comment by jbathegit** [comment]
So, and assuming I'm understanding correctly, the idea would be now reserve the first 6 bytes of the str input string for toggling on/off the chunking logic within ufbtab(), and only the remaining (up to) 74 bytes would be used to specify what mnemonics you want to read from each subset?  If so, that seems a bit kludgy to me, not to mention potentially confusing to users.  For example, what if one day there was a legitimate mnemonic which happened to have the name "PARTAB"?  Or would you always be relying on case-sensitivity to distinguish between "PARTAB" (mnemonic) and "partab" (toggle switch) inside of str?

Bottom line, I'm honestly not a fan of this approach.  However, if you wanted to do something along these same lines, then I think it might make more sense to instead focus on one of the other input parameters to ufbtab(), such as i1.  For example, and similar to how lunin already has an additional meaning when it's input as a negative number, perhaps you could now turn the chunking logic on within ufbtab() if (and only if) i1 is set to the additive inverse of its usual positive value?  To me that would be cleaner and a bit less error-prone.

Thoughts?

**Comment by jack-woollen** [comment]
First, the partab switch is one and done, in and out, no cludging with mnemonic  strings. There are other protected mnemonics like IREC and ISUB. The lower case form makes it highly unlikely that it interferes with any other purpose. Also the partab string to switch part has no other characters. In essence its very similar to how openbf uses the activity string to control behavior. But never mind, I'll fall back on your original idea of adding a subroutine next to ufbtab to switch part on and off, and make the PR. Thanks for your comments

**Comment by jack-woollen** [comment]
@jbathegit Hey Jeff, I'm setting up a PR but I noticed something odd when testing things. See the output below and notice that it came from using bufr/12.2.0.  That's funny isn't it?

![Image](https://github.com/user-attachments/assets/d3311a92-f55b-4e15-9480-0e2da996c519)

**Comment by jbathegit** [comment]
Yeah, agreed that's definitely odd.  What's interesting though is that I get the same sinv output running on that same input file in v12.1.0, v12.0.0, and even going all the way back to v11.7.0.  So whatever it is isn't unique to just the latest v12.2.0

It seems notable as well that the total number of returned reports for this particular input file (15999367) is within a hair whisker of the dimensioned limit (of 16000000) for the utility array, and which doesn't seem like it could be a coincidence.  I wonder if something in memory is getting clobbered?

**Comment by jack-woollen** [comment]
I printed out len and iret returned from getcfmng for code 225. Values below. 

  1207985152          -1

**Comment by jbathegit** [comment]
An iret of -1 from getcfmng() means it couldn't find the input icode value (in this case, 225) in the underlying master table.  And in this case that makes sense because the BUFR messages in the b012/xx024 tank look like they're being encoded using a master table version number of 23, and the values of 225 (for "NOAA 20") and 226 ("for NOAA 21") didn't become valid for SAID/0-01-007 until nearly 2 years later when version 27 of the tables was released.

But in any case, whenever you get an iret of -1 from getcfmng(), then you shouldn't necessarily rely on having any sensible corresponding values for str and len.  Specifically, in that case getcfmng() won't even write anything into len, so the resulting value of len will just be whatever happened to be stored in that memory location when the variable was first allocated by the satcode() subroutine.

So bottom line, what you're seeing for those len and iret values is what I would expect in this case, and something else may be going on here.  I'll dig some more.

**Comment by jack-woollen** [comment]
That's the answer then. The code table for version 23 doesn't contain said 225. The only problem is that value for said was coded into the dumpfile, no matter which version of bufrlib coded it. So I think getcfmng should point at the latest code/flag tables.

**Comment by jbathegit** [comment]
To follow-up, the BUFR messages in the b012/xx024 tank are compressed, which in turn means that after the current version of ufbtab() reads each successive message from that tank, it almost immediately encounters the following if block:
```
       if(iret+msub(lun)>i2) then
         overflow = .true.
         exit outer
       endif
```
which of course is there to ensure that there's enough room remaining in the tab array before it tries to add in the number of subsets  (i.e. msub(lun)) from the latest message.  So that's why it's hitting the overflow section just shy of going over the limit, because it's checking for a sufficient amount of space in tab before it stores any subsets from the latest message.

**Comment by jbathegit** [comment]
> That's the answer then. The code table for version 23 doesn't contain said 225. The only problem is that value for said was coded into the dumpfile, no matter which version of bufrlib coded it. So I think getcfmng should point at the latest code/flag tables.

I agree that it now looks like ufbtab() is behaving as expected, as opposed to something getting inadvertently clobbered in memory or other nefarious behavior.

However, I disagree that getcfmng() should always point at the latest code/flag tables.  That and other parts of the library are specifically designed to pay strict attention to the master table version number that's encoded in any message it reads in, and to in turn refer to that specific version of the table when trying to interpret any corresponding code or flag table value from that message.  The main reason for this is that such meanings could conceivably change from one version of the tables to the next, and it was never WMO's intent that any such definition would remain forever frozen once it was initially defined, or that any decoding software should only ever expect to need to look at whatever happens to be the latest version of the tables.  Again, there's no guarantee that any particular version of the master tables is always going to be a superset of all previous versions.

So if you really want sinv to have access to the meaning strings of "NOAA 20" and "NOAA 21" for this particular tank and/or dumpfile, then I would argue that whatever process is encoding that tank and/or dumpfile should be corrected to specify a tables version number of at least 27, and arguably something even more modern (i.e. the latest version from last November is version 43, and 44 is due out in another couple of months :-)  That's easy enough to do in any code with just a simple call to [pkvs01(), and with s01mnem='MTV'](https://noaa-emc.github.io/NCEPLIBS-bufr/s013vals_8F90.html#ac8fa984389323f08dd1cbd3aa6b57da9).

In any case, please understand that in this discussion we're not talking about a version of the BUFRLIB (e.g. 12.2.0, 12.1.0, etc.) which was used when compiling a program, but rather a version of the master tables (e.g. 13, 14, ..., 43) that was used when encoding a BUFR message within that program.  These are two entirely different and unrelated sets of version numbers.

**Comment by jack-woollen** [comment]
I get all that Jeff. I would add though that each bufrlib version defines its own default mtv, so there is that relationship. And, at least, the mtv default should reflect the most recent tables available in a particular bufrlib version.

**Comment by jbathegit** [comment]
Fair point that each version of BUFRLIB does indeed contain a default master table version (mtv) value to use when writing/encoding a BUFR message, and in fact if you look through the release note history you can see where I've occasionally increased that default value over the years.  For example, with v11.3.0 back in 2019, I increased the default from 13 to 29.  And then later with v11.6.0 in 2021 I upped it even further to 36.  But it's important to note that that's just the default version number to use when writing/encoding messages, whereas when reading/decoding messages (e.g. using getcfmng() within the sinv utility) it never even looks at that value of mtv.  Instead, it always uses the code/flag table corresponding to whatever version number is encoded in that particular message (and again, I would argue that's what it always should do!)

Going forward, we could certainly consider having mtv default to whatever the most recent master table version number is available at the time each new library version is released, but that still wouldn't solve the problem here unless whatever application program is writing/encoding those particular b012/xx024 tank files was then relinked to that version of the library.  Or at least to a version newer than v11.3.0, which now makes me wonder how old that particular application program really is, or how it even ended up encoding a version number of 23 in the first place when that was never one of the default values in the library?  Again, the mtv default values gradually went from 13 to 29 to 36 within the library, so I'm puzzled as to how those b012/xx024 messages are now being encoded with a version number of 23 unless that particular application program was explicitly setting them to 23 (perhaps using pkvs01()?)

In any case, none of the points in this discussion prevent us from moving forward with the chunking toggle logic that you've now added into ufbtab(), so I'll go ahead and review your recent updates to that PR.
